compiler/rustc_hir_typeck/src/generator_interior/mod.rs

```diff
@@ -426,6 +445,13 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
                 }
                 _ => intravisit::walk_expr(self, expr),
             },
+            ExprKind::Assign(_, rhs, _) => {
+                if self.task_context_hid.is_some() && matches!(rhs.kind, hir::ExprKind::Yield(..)) {
+                    self.in_await_yield = true;
+                }
+                intravisit::walk_expr(self, expr);
+                self.in_await_yield = false;
```

Comment:
Is it ever possible that this is nested, like a yield inside of a yield? If so, then we need to store the old state of `in_await_yield`. If not, we probably should do a debug (or regular) assertion to make sure it's always false as a precondition.